certora/spec/CVL2/timelockAuthorizerTed.spec

```diff
@@ -0,0 +1,560 @@
+
+    // Invoke any function
+    calldataarg args;
+    f(e, args);
+
+    uint256 delayAfter = getActionIdRevokeDelay(actionId);
+
+    assert delayBefore != delayAfter =>
+        f.selector == sig:setRevokeDelay(bytes32, uint256).selector,
+        "_revokeDelays modified by function other than setRevokeDelay.";
+    assert delayAfter != delayBefore =>
+            e.msg.sender == getTimelockExecutionHelper(),
+        "Delay changed by other entity than execution helper.";
+}
+
+
+// STATUS - verified
+// Permissions granted on target is removed when:
+// `revokePermission` is called by a revoker or the `_executionHelper`
+// or when `renouncePermission` is called (by anybody)
+// Permissions granted on target is added when `grantPermission` is
+// called by the `_executionHelper` or some granter.
+rule grantedPermissionsChangeOnlyByAllowedFunctions(env e, method f) {
+    bytes32 actionId;
+    address account;
+    address where;
+    bool grantedBefore = isPermissionGrantedOnTarget(actionId, account, where);
+
+    // Invoke any function
+    calldataarg args;
+    f(e, args);
+
+    bool grantedAfter = isPermissionGrantedOnTarget(actionId, account, where);
+
+    assert grantedBefore && !grantedAfter =>
+        f.selector == sig:revokePermission(bytes32, address, address).selector ||
+        f.selector == sig:renouncePermission(bytes32,address).selector,
+        "permission revoked by function other than revokePermission.";
+    assert !grantedBefore && grantedAfter =>
+        f.selector == sig:grantPermission(bytes32, address, address).selector,
+        "permission granted by function other than grantPermission.";
+    assert grantedBefore != grantedAfter =>
+        e.msg.sender == getTimelockExecutionHelper() ||
+        (grantedAfter && isGranter(actionId, e.msg.sender, where)) ||
+        (grantedBefore && (isRevoker(e.msg.sender, where)) || f.selector == sig:renouncePermission(bytes32,address).selector),
+        "Granted permission changed by other entity than execution helper or revoker.";
+}
+
+
+// STATUS - verified
+// This rule checks, that when a change of delay is scheduled, a new
+// ScheduledExecution is created and its executableAt meets, that
+// the waiting time for the execution to be executed is at least
+// MINIMUM_CHANGE_DELAY_EXECUTION_DELAY. Additionaly, when the delay has
+// been lowered, then the time it will take before this execution is executable
+// is at least the difference between the old delay and the new delay.
+// We assume e.block.timestamp + execution delay < max_uint256
+rule scheduleDelayChangeHasProperDelay(env e, bytes32 actionId) {
+    uint256 delayBefore = getActionIdDelay(actionId);
+    uint256 newDelay;
+    address[] executors;
+    uint256 numberOfScheduledExecutionsBefore = getSchedExeLength();
+    uint256 calculatedDelay = _getDelayChangeExecutionDelay(delayBefore, newDelay);
+
+    require(delayBefore <= MAX_DELAY());
```

Comment:
maybe nicer to require the invariant, as it shows that we have proven this assumption

Reply:
Yes, absolutely, I added `requireInvariant notGreaterThanMax(actionId);` 👍